### PR TITLE
Make # of log pipelines configurable and default to GOMAXPROCS

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?

Makes the # of logs pipelines configurable (previously hard coded to `4`) and set the default to `GOMAXPROCS` with a soft maximum of 4. 

### Motivation

Logs pipelines are inherently limited by a single CPU core due to compression overhead. The previous hardcoded setting of `4` was overfit for small hosts, and possibly underfit for large hosts (with more than 4 cores available). 

With this change we set the number of pipelines to `GOMAXPROCS` but also allow it to be customized manually by `logs_config.pipelines`. Limiting the setting to 4 ensures hosts with a lot of CPUs don't over provision too many logs pipelines but leaves room to adjust if needed. 

### Describe how to test/QA your changes

QA done by validating the logs agent still works (SMP results in PR). 
- This should have no out of the box behavior change. Agent will just better utilize available resources.

### CODE REVIEW

**Comment by gh123man** [comment]
Other teams that review this for the non-logs components: 
- Do you want to use the same # of pipelines as the logs agent all the time? or would you prefer a different default setting? (Currently using the same as the logs pipeline)

File: pkg/security/reporter/reporter.go
Comment (paulcacheux) [inline pkg/security/reporter/reporter.go:55]: This code path can be called from the system probe, and should probably take the config itself as an input, is there a default you can use in this PR and we can take care of making it configurable in a later PR.

I'm also not a huge fan of having a `logs_config` option configuring some CWS related code.

Comment (paulcacheux) [reply pkg/security/reporter/reporter.go:55]: related to https://github.com/DataDog/datadog-agent/pull/31190#issuecomment-2483934221, I think a default would be best for now (either 4 or rutime.GOMAXPROCS)

Comment (gh123man) [reply pkg/security/reporter/reporter.go]: Thanks for the feedback - I defaulted it to `GOMAXPROCS` and avoided the logs agent config.

Comment (gh123man) [reply pkg/security/reporter/reporter.go]: I am not super aware of your use case. But FWIW - in the future, unless you are sending an extremely high volume (10+ MB/sec) of data you probably don't need more than 1 pipeline which could same some memory overhead.

Comment (paulcacheux) [reply pkg/security/reporter/reporter.go]: Yeah we are definitely not sending >10MB/sec, IMO your default is good for now, and you can do the same in `pkg/compliance`. I will do some benchmarks/measurements once this is merged to see if 1 is better. Thanks a lot !

**Review by paulcacheux — APPROVED** [review]

**Review by vickenty — APPROVED** [review]

File: pkg/config/setup/config.go
Comment (sgnn7) [inline pkg/config/setup/config.go:1559]: @gh123man Would this have similar issues as the count of forwarders on high core-count servers (e.g. 96+)? IIRC, we generally have ended up needing to do logarithmic scaling with core counts in other parallelized code (workers, forwarders) to avoid problems.

Comment (gh123man) [reply pkg/config/setup/config.go:1559]: Been chatting with @scottopell about this. 
I am going to run some benchmarks (idle agent collecting no logs with a large number of pipelines) and compare it to the baseline/current implementation to see what the expected overhead is.

**Review by sgnn7 — APPROVED** [review]
@gh123man One small nit but not a blocker

**Review by paulcacheux — APPROVED** [review]

**Comment by remeh** [comment]
This PR will double the amount of (possible idle) connections to the intake, let's make sure it's fine, but otherwise, LGTM!

File: comp/otelcol/logsagentpipeline/logsagentpipelineimpl/agent.go
Comment (gh123man) [reply comp/otelcol/logsagentpipeline/logsagentpipelineimpl/agent.go:213]: If the collector uses the same default configurations as the agent uses - yes it will use the default value here. The default was just moved to the config instead of being hard coded so a user could configure it if they wanted. 

To be safe (if you want) - I could return this line back to the original default (`4`) as I did for the security and compliance agent.

**Comment by jackgopack4** [comment]
is this still happening?

**Review by jackgopack4 — APPROVED** [review]

**Comment by gh123man** [comment]
@jackgopack4  Yep! Was blocked by the merge freeze - thanks for the signoff!

**Pull request merged**
